### ISSUE TITLE
IPFS support

### ISSUE THREAD

**Issue by ShalokShalom** [netlify/netlify-cms#502]
IPFS is a distributed peer-to-peer network, which offers the hosting of static sites:

https://ipfs.io/
https://ipfs.io/ipfs/QmdPtC3T7Kcu9iJg6hYzLBWR5XCDcYMY7HV685E3kH3EcS/2015/09/15/hosting-a-website-on-ipfs/

Can you support it?

**Comment by Benaiah** [comment]
Editing a site which is hosted on IPFS should be perfectly feasible once you have a build process which deploys to IPFS - Netlify CMS doesn't handle building or hosting the site whatsoever, just editing the content. You should be able to set up a GitHub repo for your content and deploy the content to IPFS from GitHub however you want. The only missing piece you'd need is an auth server for authenticating with GitHub. We provide that for free for sites hosted on Netlify, but custom auth servers exist - see https://www.netlifycms.org/docs/custom-authentication for more details.

Using IPFS as a backend for your _source_ files, which would require direct interaction with IPFS by the CMS, would be significantly more complicated. I don't know enough about IPFS yet to say how that would be best implemented, but unless IPFS supports pushing changes directly from the browser somehow it would require an intermediary server to push changes. If that's the case, I don't see it as terribly useful to have IPFS-specific handling in the CMS, since there are other potential backends that could be used to do the same thing - for instance, if local filesystem or local git repo backends were in place, you could run the CMS on a server which pushes changes to IPFS and 

If IPFS _does_ support pushing changes from the browser, it would probably be feasible to support it. I'd be thrilled at any PRs building CMS support for it if that's the case.

**Comment by ShalokShalom** [comment]
There is also a JS implementation, which runs without any installements or proxy like it is currently essential with the go implementation. https://github.com/ipfs/js-ipfs

Plus, a member said "its just the commit workflow that needs changing"

**Comment by ShalokShalom** [comment]
Is anybody interested in this?

**Comment by chrisschaub** [comment]
This would be a great addition, a "publish to ipfs" option. I would let github host the source, no point in re-inventing that wheel. IPFS seems like a good place for the released for finished product. And yes, very interested in this.

**Comment by ShalokShalom** [comment]
You could still host it on any source hosting platform of your choice, such as Gitea. Choice is always nice and its not to hard to follow the standard that git is. No need to make it Github specific for no reason, I guess. :smiley:

**Comment by hsn10** [comment]
Yes, I would like to have ability to publish into IPFS as well. Build process should support updating _dnslink as well. There is no need for netlify to run IPFS gateway.

**Comment by mfcodeworks** [comment]
+1 for this, not sure if it's in the pipeline but I'd love an option to publish to IPFS, maybe including IPNS for a static URL

**Comment by jikkujose** [comment]
This would be awesome!

**Comment by decebal** [comment]
+1

**Comment by woodcock3** [comment]
Yes iPad support would be great. Maybe net lift could also offer a pinning service like [pinata](https://www.pinata.cloud/). If there is a build process to include netlifycms that would be awesome.

**Comment by chrisschaub** [comment]
I am working on this for a client, I'll see what I can post. But as a starter, this exists:

https://github.com/ipfs-shipyard/ipfs-deploy

So once you get a "build" you can run this and it will deploy to an ipfs pinning service (supports Pinata etc.) and then it will update DNS for you! Every bulid to IPFS requires you to update your DNS, ENS, HNS etc... so that it knows where to find the new IPFS data.

This plugin, for Gatsby, doesn't really work for Gatsby v4 .. https://www.gatsbyjs.com/plugins/gatsby-plugin-ipfs/ . It is a key piece so the built files have relative paths.

**Comment by christophkoenekamp** [comment]
+1
